chrome/content/zotero/elements/itemPane.js

```diff
@@ -0,0 +1,522 @@
+
+			this._container.addEventListener("keypress", this._handleKeypress);
+			this._paneParent.addEventListener('scroll', this._handleContainerScroll);
+
+			this._paneHiddenOb = new MutationObserver(this._handlePaneStatus);
+			this._paneHiddenOb.observe(this._paneParent, {
+				attributes: true,
+				attributeFilter: ["hidden"],
+				subtree: true,
+			});
+			this._initIntersectionObserver();
+
+			this._unregisterID = Zotero.Notifier.registerObserver(this, ['item'], 'itemPane');
+
+			this._disableScrollHandler = false;
+			this._pinnedPaneMinScrollHeight = 0;
+		}
+
+		destroy() {
+			this._container.removeEventListener("keypress", this._handleKeypress);
+			this._paneParent.removeEventListener('scroll', this._handleContainerScroll);
+
+			this._paneHiddenOb.disconnect();
+			this._intersectionOb.disconnect();
+
+			Zotero.Notifier.unregisterObserver(this._unregisterID);
+		}
+
+		async render() {
+			if (!this.initialized || !this.item) {
+				return;
+			}
+			let item = this.item;
+			Zotero.debug('Viewing item');
+			this._isRendering = true;
+
+			let panes = this.getPanes();
+			let pendingBoxes = [];
+			let inTrash = ZoteroPane.collectionsView.selectedTreeRow && ZoteroPane.collectionsView.selectedTreeRow.isTrash();
+			for (let box of [this._header, ...panes]) {
+				if (!box.showInFeeds && item.isFeedItem) {
+					box.style.display = 'none';
+					box.hidden = true;
+					continue;
+				}
+				else {
+					box.style.display = '';
+					box.hidden = false;
+				}
+				
+				if (this.mode) {
+					box.mode = this.mode;
+					
+					if (box.mode == 'view') {
+						box.hideEmptyFields = true;
+					}
+				}
+				else {
+					box.mode = 'edit';
+				}
+				
+				box.item = item;
+				box.inTrash = inTrash;
+				// Render sync boxes immediately
+				if (!box.hidden && box.render) {
+					if (box.render instanceof AsyncFunction) {
+						pendingBoxes.push(box);
+					}
+					else {
+						box.render();
+					}
+				}
+			}
+
+			let pinnedPaneElem = this.getPane(this.pinnedPane);
+			let pinnedIndex = panes.indexOf(pinnedPaneElem);
+			
+			this._paneParent.style.paddingBottom = '';
+			await waitFrames(3);
+			if (pinnedPaneElem) {
+				let paneID = pinnedPaneElem.dataset.pane;
+				this.scrollToPane(paneID, 'instant');
+				this.pinnedPane = paneID;
+			}
+			else {
+				this._paneParent.scrollTo(0, 0);
+			}
+
+			// Only render visible panes
+			for (let box of pendingBoxes) {
+				if (pinnedIndex > -1 && panes.indexOf(box) < pinnedIndex) {
+					continue;
+				}
+				if (!this.isPaneVisible(box.dataset.pane)) {
+					continue;
+				}
+				await box.render();
+			}
+			// After all panes finish first rendering, try secondary rendering
+			for (let box of panes) {
+				if (!box.secondaryRender) {
+					continue;
+				}
+				if (pinnedIndex > -1 && panes.indexOf(box) < pinnedIndex) {
+					continue;
+				}
+				if (this.isPaneVisible(box.dataset.pane)) {
+					continue;
+				}
+				await box.secondaryRender();
```

Comment:
We also allow a `secondaryRender`, which allows you to roughly render some initial content/set a box height in sync, and then render real contents that have to go in async later when needed. Would be useful for plugin API.

The setting data and rendering steps are separately controlled by the CE and plugins can lazy-load their panes to minimize resource usage (so does our attachment preview).

Reply:
Neat. I guess one question is…should plugins be able to render anything synchronously, or just set a box height?

Reply:
If the rendering is fast, e.g. display some text, fine, I guess? We could have some warnings for sync render functions that take too long to recommend them to use async.